File: tutorials/golang/06-hello-structured-logging/logger.go
package main

import (
	"io"
	"log/slog"
)

// NewJSONLogger creates a structured JSON logger writing to w.
func NewJSONLogger(w io.Writer) *slog.Logger {
	return slog.New(slog.NewJSONHandler(w, &slog.HandlerOptions{
		Level: slog.LevelDebug,
	}))
}

// LogRequest logs an HTTP-like request event.
func LogRequest(logger *slog.Logger, method, path string, status int, durationMs float64) {
	logger.Info("request",
		slog.String("method", method),
		slog.String("path", path),
		slog.Int("status", status),
		slog.Float64("duration_ms", durationMs),
	)
}

// LogError logs an error event.
func LogError(logger *slog.Logger, msg string, err error) {
	logger.Error(msg,
		slog.String("error", err.Error()),
	)
}

// LogStartup logs a startup event with service metadata.
func LogStartup(logger *slog.Logger, service, version, addr string) {
	logger.Info("startup",
		slog.String("service", service),
		slog.String("version", version),
		slog.String("addr", addr),
	)
}

File: tutorials/lua/06-hello-structured-logging/src/logger.lua
local json = require("dkjson")

local M = {}

function M.log(level, message, metadata)
    local entry = {
        level = level,
        message = message,
        timestamp = os.date("!%Y-%m-%dT%H:%M:%SZ")
    }
    if metadata then
        for k, v in pairs(metadata) do
            entry[k] = v
        end
    end
    return json.encode(entry)
end

function M.info(message, metadata)  return M.log("info", message, metadata) end
function M.warn(message, metadata)  return M.log("warn", message, metadata) end
function M.error(message, metadata) return M.log("error", message, metadata) end

return M
